### PR TITLE
Performance tests demo

### PR DESCRIPTION
**Pull request by Blast545**
Testing performance test usage for `rcutils/error_handling.h` and `rcutils/logging.h`. 

IMO it would be a good idea to combine coverage measurements to see which parts of the code are run with the performance tests, investigation pending.

### CODE REVIEW

**Review by fujitatomoya — COMMENTED** [review]
It would be nice to integrate performance benchmark test. but for doing that, we should be able to have some criteria or threshold to decide if performance is not good enough? I think that most likely this performance check is gonna be useful to make sure if the staging PR does not affect the performance aspect. Could you share your thoughts about this kind of plan?

**Comment by Blast545** [comment]
@fujitatomoya There is a pending discussion on #278 for this package, we will be discussing what's reasonable to test for each package in terms of performance and include checks on CI to be able of checking when a PR affects considerably performance to be considered before merging.

File: CMakeLists.txt [hunk 1]
```diff
@@ -145,6 +145,12 @@ if(BUILD_TESTING)
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14")
   endif()
 
+  # BENCHMARK_F seems to break cppcheck 2.1
+  # set(ament_cmake_cppcheck_FOUND TRUE)
```
Comment (wjwwood) [inline CMakeLists.txt:149]: Is this just disabling cppcheck? I don't think that's an ok solution. If it is only disabling it for the tests somehow that would be ok I think. You can also disable cppcheck locally, e.g.:

https://github.com/ros2/rclcpp/blob/c214ddc57b1f2e75076aa087e77c9a06b769008d/rclcpp/include/rclcpp/node.hpp#L195

**Review by wjwwood — APPROVED** [review]

File: CMakeLists.txt [hunk 2]
```diff
@@ -505,6 +511,15 @@ if(BUILD_TESTING)
...
     test/test_macros.cpp
   )
+
+add_performance_test(benchmark_test_logging test/test_benchmark_logging.cpp)
+target_link_libraries(benchmark_test_logging ${PROJECT_NAME})
+target_include_directories(benchmark_test_logging PUBLIC include)
+
+add_performance_test(benchmark_test_err_handle test/test_benchmark_error_handling.cpp)
+target_link_libraries(benchmark_test_err_handle ${PROJECT_NAME})
+target_include_directories(benchmark_test_err_handle PUBLIC include)
```
Comment (ahcorde) [inline CMakeLists.txt:521]:
```suggestion
  add_performance_test(benchmark_test_logging test/test_benchmark_logging.cpp)
  target_link_libraries(benchmark_test_logging ${PROJECT_NAME})
  target_include_directories(benchmark_test_logging PUBLIC include)

  add_performance_test(benchmark_test_err_handle test/test_benchmark_error_handling.cpp)
  target_link_libraries(benchmark_test_err_handle ${PROJECT_NAME})
  target_include_directories(benchmark_test_err_handle PUBLIC include)
```

File: test/test_benchmark_logging.cpp [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+{
+  const rcutils_log_location_t * location;
+  int level;
+  std::string name;
+  rcutils_time_point_value_t timestamp;
+  std::string message;
+};
+LogEvent g_last_log_event;
+
+static void simple_benchmark(benchmark::State & state) {
```
Comment (ahcorde) [inline test/test_benchmark_logging.cpp:43]: rename test

```suggestion
static void benchmark_logging(benchmark::State & state) {
```

File: test/test_benchmark_logging.cpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    rcutils_log(NULL, RCUTILS_LOG_SEVERITY_ERROR, "", "%s", "");
+    rcutils_log(NULL, RCUTILS_LOG_SEVERITY_FATAL, NULL, "%s", "");
+
+    // restore original state
+    rcutils_logging_set_default_logger_level(original_level);
+    rcutils_logging_set_output_handler(original_function);
+  }
+}
+
+BENCHMARK(simple_benchmark);
```
Comment (ahcorde) [inline test/test_benchmark_logging.cpp:88]:
```suggestion
BENCHMARK(benchmark_logging);
```

**Review by ahcorde — DISMISSED** [review]
Does it make sense to add a new folder for the benchmark tests?

File: test/test_benchmark_logging.cpp (hunk 4 above)
Comment (Blast545) [reply test/test_benchmark_logging.cpp]: See 99818c0

File: test/test_benchmark_logging.cpp (hunk 3 above)
Comment (Blast545) [reply test/test_benchmark_logging.cpp]: See 99818c0

File: CMakeLists.txt (hunk 2 above)
Comment (Blast545) [reply CMakeLists.txt]: See 87e22e0

File: CMakeLists.txt (hunk 1 above)
Comment (Blast545) [reply CMakeLists.txt]: This was to disable cppcheck, I copied it from the performance fixture test demo. Although I enabled it locally and it seems to be working fine. I'll run CI and check if `cppcheck` causes problems with this PR.

**Comment by Blast545** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12307/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7278/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10022/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12204/)

**Review by ahcorde — COMMENTED** [review]
I have two suggestions: 
 - call the target `benchmark_logging` instead of `benchmark_test_logging`. I think is redundant somehow. 
 - remove the word `test` from the filename.  (test_benchmark_logging.cpp -> benchmark_logging.cpp)

**Comment by Blast545** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12376/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7345/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10088/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12288/)

**Comment by clalancette** [comment]
@Blast545 What's the status of this one?  Should we merge it?

**Comment by ahcorde** [comment]
@clalancette I believe that rhe idea it's to create end-to-end performance tests instead of microbenchmarks.

Maybe @chapulina can add more context here.

**Comment by chapulina** [comment]
Yes, the new plan is to create end-to-end tests. But it looks like this PR is already approved and close to getting in, so it would be a shame to throw it away. What's missing to merge it?

**Comment by clalancette** [comment]
> Yes, the new plan is to create end-to-end tests. But it looks like this PR is already approved and close to getting in, so it would be a shame to throw it away. What's missing to merge it?

Yeah, agreed.  If this is almost ready to go, then I would say we get this in, but don't build any *new* micro-benchmarks.

File: CMakeLists.txt [hunk 5]
```diff
@@ -145,6 +145,9 @@ if(BUILD_TESTING)
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14")
   endif()
 
+  find_package(ament_cmake_google_benchmark REQUIRED)
```
Comment (cottsay) [inline CMakeLists.txt:148]: You should be able to drop this, as well as the reference in the `package.xml`. The `performance_test_fixture` dependency should be sufficient for what you're doing here.

File: CMakeLists.txt [hunk 6]
```diff
@@ -505,6 +508,15 @@ if(BUILD_TESTING)
   rcutils_custom_add_gtest(test_macros
     test/test_macros.cpp
   )
+
+  add_performance_test(benchmark_logging test/benchmark/benchmark_logging.cpp)
+  target_link_libraries(benchmark_logging ${PROJECT_NAME})
+  target_include_directories(benchmark_logging PUBLIC include)
```
Comment (cottsay) [inline CMakeLists.txt:514]: These should get the same `if(TARGET ...)` treatment that the gtests are getting above. If `benchmark` isn't available, the tests won't be built, and these lines will fail (the same as gtest).

**Review by cottsay — CHANGES_REQUESTED** [review]

**Comment by Blast545** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13427/)

File: CMakeLists.txt (hunk 6 above)
Comment (Blast545) [reply CMakeLists.txt]: Sure, see b830dba

File: CMakeLists.txt (hunk 5 above)
Comment (Blast545) [reply CMakeLists.txt]: Thanks for the advice, addressed with b830dba

**Comment by clalancette** [comment]
Here is full CI on this to make sure everything is happy.  If it is, I'll go ahead and merge it:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13893/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8744/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11602/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13984/)

**Review by clalancette — APPROVED** [review]

**Pull request merged**